dashboard/pages/data_projection/callbacks.py

```diff
@@ -0,0 +1,306 @@
+import base64
+import functools
+import io
+import uuid
+from datetime import datetime
+
+import pandas as pd
+import plotly.graph_objects as go
+import pyarrow as pa
+from dash import Input, Output, State, callback, dcc, html, no_update
+from sklearn.decomposition import PCA
+from umap import UMAP
+
+from dashboard.data.controls import generate_controls, controls_index_annotator
+from dashboard.data.preprocess import MergedAssaysPreprocessor
+from dashboard.data.utils import eos_to_ecbd_link
+from dashboard.pages.components import make_file_list_component
+from dashboard.storage import FileStorage
+from dashboard.visualization.plots import plot_projection_2d, make_projection_plot
+from dashboard.visualization.text_tables import pca_summary, table_from_df
+
+PROJECTION_SETUP = [
+    (PCA(n_components=2), "PCA"),
+    (
+        UMAP(
+            n_components=2,
+            n_neighbors=10,
+            min_dist=0.1,
+        ),
+        "UMAP",
+    ),
+]
+
+# === STAGE 1 ===
+
+
+def on_projection_files_upload(
+    content: str | None,
+    filenames: list[str],
+    last_modified: int,
+    stored_uuid: str,
+    file_storage: FileStorage,
+) -> html.Div:
```

Comment:
the initial callback for each process has to set `stored_uuid` if it does not exists (that happens whenever the user simply does not have the `stored_uuid` set but wishes to use the process), because without it the processing will fail

As such you have to return the `stored_uuid` as well from this callback and save it to appropriate uuid holding `State`

Reply:
Ahh, okay I see, thanks for the explanation, it's corrected now 👍